components/camera/rtsp/rtsp.go

```diff
@@ -0,0 +1,281 @@
+	}
+	client := &gortsplib.Client{}
+	// connect to the server - be sure to close it if setup fails.
+	var clientSuccessful bool
+	var clientErr error
+	err = client.Start(u.Scheme, u.Host)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if !clientSuccessful {
+			clientErr = client.Close()
+		}
+	}()
+	// get MJPEG decodings
+	mjpegFormat, mjpegDecoder := mjpegDecoding()
+
+	// find published tracks, and MJPEG specifically
+	tracks, baseURL, _, err := client.Describe(u)
+	if err != nil {
+		return nil, multierr.Combine(err, clientErr)
+	}
+	track := tracks.FindFormat(&mjpegFormat)
+	if track == nil {
+		err := errors.Errorf("MJPEG track not found in rtsp camera %s", u)
+		return nil, multierr.Combine(err, clientErr)
+	}
+	// Setup the MJPEG track
+	_, err = client.Setup(track, baseURL, 0, 0)
+	if err != nil {
+		return nil, multierr.Combine(err, clientErr)
+	}
+	// define the camera
+	rtspCam := &rtspCamera{
+		u:      u,
+		logger: logger,
+	}
+	var gotFirstFrameOnce bool
+	gotFirstFrame := make(chan struct{})
+	// On packet retreival, turn it into an image, and store it in shared memory
+	client.OnPacketRTP(track, mjpegFormat, func(pkt *rtp.Packet) {
+		img, err := mjpegDecoder(pkt)
+		if err != nil {
+			return
+		}
+		// wait for a frame
+		if img == nil {
+			return
+		}
+		rtspCam.latestFrame.Store(img)
+		if !gotFirstFrameOnce {
+			close(gotFirstFrame)
+			gotFirstFrameOnce = true
+		}
+	})
+	// play the track
+	_, err = client.Play(nil)
+	if err != nil {
+		return nil, multierr.Combine(err, clientErr)
+	}
+	clientSuccessful = true
+	// read the image from shared memory when it is requested
+	cancelCtx, cancel := context.WithCancel(context.Background())
+	reader := gostream.VideoReaderFunc(func(ctx context.Context) (image.Image, func(), error) {
+		select {
+		case <-cancelCtx.Done():
+			return nil, nil, cancelCtx.Err()
+		case <-ctx.Done():
+			return nil, nil, ctx.Err()
+		default:
+		}
+		<-gotFirstFrame // block until you get the first frame
+		return rtspCam.latestFrame.Load().(image.Image), func() {}, nil
+	})
+	// define and return the camera
+	rtspCam.VideoReader = reader
+	rtspCam.client = client
+	rtspCam.cancelCtx = cancelCtx
+	rtspCam.cancelFunc = cancel
+	cameraModel := &transform.PinholeCameraModel{attrs.IntrinsicParams, attrs.DistortionParams}
+	rtspCam.clientReconnectBackgroundWorker()
+	return camera.NewFromReader(ctx, rtspCam, cameraModel, camera.ColorStream)
+}
+
+func restartClient(rc *rtspCamera) error {
+	err := rc.client.Close()
+	if err != nil {
+		rc.logger.Debugf("error closing while trying to restart client: %s", err)
+	}
+	// replace the client with a new one
+	client := &gortsplib.Client{}
+	rc.client = client
+	// restart the client
+	err = rc.client.Start(rc.u.Scheme, rc.u.Host)
+	if err != nil {
+		return err
+	}
+	// get MJPEG decodings
+	mjpegFormat, mjpegDecoder := mjpegDecoding()
+
+	// find published tracks, and MJPEG specifically
+	tracks, baseURL, _, err := rc.client.Describe(rc.u)
```

Comment:
can we find a way to refactor all of the resetup code?

Reply:
created "ReconnectClient" method